### PR TITLE
Extraplbfgs2

### PR DESCRIPTION
**Pull request by cortner**
This is small modification to LBFGS which should leave the default behaviour completely unchanged. But for some problems I get significant speed-up even over the HZ-CG.

```
--------------------
Rosenbrock Example:
--------------------
LBFGS Default Options: g_calls = 157, f_calls = 157
CG Default Options: g_calls = 65, f_calls = 86
LBFGS + Backtracking + Extrapolation: g_calls = 23, f_calls = 52
--------------------------------------
p-Laplacian Example (preconditioned):
--------------------------------------
LBFGS Default Options: g_calls = 15, f_calls = 15
CG Default Options: g_calls = 8, f_calls = 11
LBFGS + Backtracking + Extrapolation: g_calls = 9, f_calls = 18
--------------------------------------
JuLIP Example (preconditioned):
--------------------------------------
LBFGS Default Options: g_calls = 168, f_calls = 168
CG Default Options: g_calls = 93, f_calls = 135
LBFGS + Backtracking + Extrapolation: g_calls = 47, f_calls = 124
```

(only the first two of these tests are contained in this PR)

The modification adds a simple step-size extrapolation step to LBFGS, which reverts to 1.0 when safe to do so.

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -7,30 +7,31 @@ using Base.Test
...
-    "api.jl",
-    "golden_section.jl",
-    "brent.jl",
-    "type_stability.jl",
-    "array.jl",
-    "constrained.jl",
-    "callbacks.jl",
-    "precon.jl",
-    "initial_convergence.jl"
+   #  "types.jl",
```
Comment (anriseth) [inline test/runtests.jl]: You've commented out all the other tests.

Comment (pkofod) [inline test/runtests.jl]: Can't seem to reply to review comments but see #259 for benchmark related discussions. I intend to move it to a package and maybe leverage pkgbenchmarks.jl

**Comment by cortner** [comment]
> Improvements are great. Are there any references in the literature on this approach?
Yes, as in this is one of the standard extrapolation techniques from N/W 
No, as in I don't think anybody in optimisation recommends this for BFGS (but I could be wrong) 

> I think it could be good for future performance improvements, if we create a larger set of benchmark problems (test them in Cutest?) so that can make decisions on default behavior the algorithms.

That would be great.

**Comment by cortner** [comment]
I should also say: this is a stop-gap solution; in the future this should be re-organised; see #294 .

**Comment by pkofod** [comment]
I see my comment didn't make it. But I hope to implement #259 in a separate package soon. Then it shouldn't be more than running the benchmarks there.

**Comment by cortner** [comment]
any idea why the linux test doesn't pass?

**Comment by cortner** [comment]
I just noticed that each iteration there is still one additional function call. This means that my benchmarks in the beginning should actually read 
```
--------------------
Rosenbrock Example:
--------------------
LBFGS Default Options: g_calls = 157, f_calls = 157
CG Default Options: g_calls = 65, f_calls = 86
LBFGS + Backtracking + Extrapolation: g_calls = 23, f_calls = 29
--------------------------------------
p-Laplacian Example (preconditioned):
--------------------------------------
LBFGS Default Options: g_calls = 15, f_calls = 15
CG Default Options: g_calls = 8, f_calls = 11
LBFGS + Backtracking + Extrapolation: g_calls = 9, f_calls = 9
--------------------------------------
JuLIP Example (preconditioned):
--------------------------------------
LBFGS Default Options: g_calls = 168, f_calls = 168
CG Default Options: g_calls = 93, f_calls = 135
LBFGS + Backtracking + Extrapolation: g_calls = 47, f_calls = 77
```

This is because at the function value from the final line-search iterate is not passed back to the main loop (in `update_g!` I think). #282 discusses this, but I wonder what a stop-gap solution would be?

**Comment by anriseth** [comment]
I just had a look at the "Initial step length" section in Nocedal and Wright.
They say that "For Newton and quasi-Newton methods, the step $\alpha_0=1$ should always be used as the initial trial step length". Have you tried to compare the performance you see here with that method? I would guess that when the quasi-Newton has a good approximation of the Hessian, this would perform better. 

The other two step length guesses that they mention (including the one implemented here), are said to be important for conjugate gradient and gradient descent. 

More flexibility in choice is of course best here - so as you say, separating this out from the specific algorithm will help us experiment with the best default options (#294).

**Comment by cortner** [comment]
> I just had a look at the "Initial step length" section in Nocedal and Wright.
They say that "For Newton and quasi-Newton methods, the step $\alpha_0=1$ should always be used as the initial trial step length".

The current LBFGS doesn't do that at all. Mine actually does: when the step choice is close to 1, then it "snaps" onto 1. The HZ line-search gets away with it because (a) it uses Wolfe and (b) it becomes more and more accurate with each iteration hence ought to approach 1 in the limit.

>  Have you tried to compare the performance you see here with that method? I would guess that when the quasi-Newton has a good approximation of the Hessian, this would perform better.

This is almost never the case in the pre-asymptotic regime. Anyhow, here are some tests in which I now always set `alphaguess = 1.0` unless the extrapolation is turned on. In the first two examples the performance is comparable (these are very easy problems) but in the last one it gets slightly worse.
```
INFO: Testing Optim
Running tests:
 * extrapolate.jl
--------------------
Rosenbrock Example:
--------------------
LBFGS Default Options: g_calls = 69, f_calls = 69
CG Default Options: g_calls = 65, f_calls = 86
LBFGS + Backtracking + Extrapolation: g_calls = 23, f_calls = 51
LBFGS + Backtracking: g_calls = 22, f_calls = 53
--------------------------------------
p-Laplacian Example (preconditioned):
--------------------------------------
LBFGS Default Options: g_calls = 17, f_calls = 17
CG Default Options: g_calls = 8, f_calls = 11
LBFGS + Backtracking + Extrapolation: g_calls = 11, f_calls = 22
LBFGS + Backtracking: g_calls = 9, f_calls = 18
--------------------------------------
JuLIP Example (preconditioned):
--------------------------------------
LBFGS Default Options: g_calls = 200, f_calls = 200
CG Default Options: g_calls = 99, f_calls = 143
LBFGS + Backtracking + Extrapolation: g_calls = 50, f_calls = 131
LBFGS + Backtracking: g_calls = 55, f_calls = 190
CG + Backtracking:  g_calls = 47, f_calls = 142
INFO: Optim tests passed
```

**Comment by cortner** [comment]
For what it's worth, if we can redesign `LineSearches` according to [LineSearches.jl #9](https://github.com/anriseth/LineSearches.jl/issues/9) then it will be easy to set various options. 

As I said above: this is a stop-gap solution, and it actually *doesn't change default behaviour*

**Comment by cortner** [comment]
any reason why this can't be merged?

**Comment by pkofod** [comment]
> any reason why this can't be merged?

Yes, I haven't had time. Not a good reason, but one nonetheless :)

There's just one thing I'm not to sure about, but I'll mention it at the line.

File: src/l_bfgs.jl [hunk 2]
```diff
@@ -146,10 +149,26 @@ function update_state!{T}(d, state::LBFGSState{T}, method::LBFGS)
     LineSearches.clear!(state.lsr)
     push!(state.lsr, zero(T), state.f_x, dphi0)
 
+    # compute an initial guess for the linesearch based on
+    # Nocedal/Wright, 2nd ed, (3.60)
+    # TODO: this is a temporary fix, but should eventually be split off into
+    #       a separate type and possibly live in LineSearches; see #294
+    if method.extrapolate && state.pseudo_iteration > 1
+        alphaguess = 2.0 * (state.f_x - state.f_x_previous) / dphi0
+        alphaguess = max(alphaguess, state.alpha/4.0)  # not too much reduction
```
Comment (pkofod) [inline src/l_bfgs.jl]: What does this line do exactly? Where does the `4.0` come from?

**Comment by cortner** [comment]
thank you - much appreciated.

**Pull request merged**